src/platform/linux/input.cpp

```diff
@@ -1583,10 +1582,51 @@ namespace platf {
    * @param input The client-specific input context.
    * @param touch_port The current viewport for translating to screen coordinates.
    * @param touch The touch event.
+   * @param input_ctx The global input context.
    */
   void
-  touch(client_input_t *input, const touch_port_t &touch_port, const touch_input_t &touch) {
-    // Unimplemented feature - platform_caps::pen_touch
+  touch(client_input_t *input, const touch_port_t &touch_port, const touch_input_t &touch, input_t &input_ctx) {
+
+    int id;
+
+    auto touchscreen = ((input_raw_t *) input_dev.get())->touch_input.get();
+    if (!touchscreen) {
+      return;
+    }
+
+
+    switch (touch.eventType)  {
+    case LI_TOUCH_EVENT_UP:
+          id = -1;
+        break;
+    case LI_TOUCH_EVENT_CANCEL:
+          id = -1;
+        break;
+    case LI_TOUCH_EVENT_CANCEL_ALL :
+        for (int i; i < 10; i++) {
+          libevdev_uinput_write_event(touchscreen, EV_ABS, ABS_MT_SLOT, i);
+          libevdev_uinput_write_event(touchscreen, EV_ABS, ABS_MT_TRACKING_ID, -1);
+        }
+        libevdev_uinput_write_event(touchscreen, EV_SYN, SYN_REPORT, 0);
+        return;
+    default:
+        id = touch.pointerId + 4;
```

Comment:
Why the `+ 4`?

Reply:
Good catch, I forgot to remove the + 4 while I was cleaning up. I was testing a theory I had about why multitouch didn't work in the first version of this PR